client/pkg/session/session.go

```diff
@@ -0,0 +1,176 @@
+package session
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"os"
+	"sync/atomic"
+
+	"github.com/wandb/wandb/client/internal/connection"
+	"github.com/wandb/wandb/client/internal/launcher"
+	"github.com/wandb/wandb/core/pkg/observability"
+	"github.com/wandb/wandb/core/pkg/service"
+)
+
+const (
+	localHost = "127.0.0.1"
+)
+
+// SessionParams holds parameters for creating a new Session
+type SessionParams struct {
+	CorePath string
+}
+
+// Session manages the lifecycle of a connection session
+type Session struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	logger     *observability.CoreLogger
+	loggerFile *os.File
+
+	// started is an atomic boolean that indicates whether the session has been
+	// started
+	started *atomic.Bool
+
+	// corePath is the path to the core binary that the session will launch
+	corePath string
+
+	// internalProcess is the launcher that manages the wandb-core process
+	internalProcess *launcher.Launcher
+
+	// address is the address of the server that the session is connected to
+	address string
+}
+
+// New creates a new Session with the provided parameters
+func New(params *SessionParams) *Session {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	s := &Session{
+		ctx:             ctx,
+		cancel:          cancel,
+		corePath:        params.CorePath,
+		started:         &atomic.Bool{},
+		internalProcess: launcher.New(),
+	}
+	s.setupLogger()
+
+	return s
+}
+
+// Address returns the address of the wandb-core server that the session is
+// connected to
+func (s *Session) Address() string {
+	return s.address
+}
+
+// LogFileName returns the name of the default log file
+func (s *Session) LogFileName() string {
+	if s.loggerFile == nil {
+		return ""
+	}
+	return s.loggerFile.Name()
+}
+
+// setupLogger sets up the default logger for the session
+func (s *Session) setupLogger() {
+	if file, _ := observability.GetLoggerPath("client"); file != nil {
```

Comment:
invert if statement to reduce nesting

`if x { doLotsOfStuff() }` => `if !x { return } doLotsOfStuff()`